x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java

```diff
@@ -2067,6 +2209,37 @@ private Cipher getDecryptionCipher(byte[] iv, String password, byte[] salt) thro
         return cipher;
     }
 
+    // public to be used in tests
+    public Tuple<byte[], byte[]> getRandomTokenBytes(boolean includeRefreshToken) {
+        return getRandomTokenBytes(getTokenVersionCompatibility(), includeRefreshToken);
+    }
+
+    Tuple<byte[], byte[]> getRandomTokenBytes(TransportVersion version, boolean includeRefreshToken) {
+        if (version.onOrAfter(VERSION_GET_TOKEN_DOC_FOR_REFRESH)) {
+            byte[] accessTokenBytes = getRandomBytes(RAW_TOKEN_BYTES_LENGTH + RAW_TOKEN_DOC_ID_BYTES_LENGTH);
+            if (includeRefreshToken) {
+                byte[] refreshTokenBytes = getRandomBytes(RAW_TOKEN_BYTES_LENGTH + RAW_TOKEN_DOC_ID_BYTES_LENGTH);
+                // an access and refresh token pair are "related" by having a common fixed-length suffix
+                System.arraycopy(
+                    accessTokenBytes,
+                    RAW_TOKEN_BYTES_LENGTH,
+                    refreshTokenBytes,
+                    RAW_TOKEN_BYTES_LENGTH,
+                    RAW_TOKEN_DOC_ID_BYTES_LENGTH
```

Comment:
Indeed I favored simplicity. I'll raise the one that does the data copying. We'll discuss over these.

Reply:
I've pushed [c7f774f](https://github.com/elastic/elasticsearch/pull/97395/commits/c7f774f8e42ae7a794ab541e97bf2e0bd78c74a0) . It looks like this option is equally simple.

Looks like we're left with a choice between:
 * throwing away 8 securely generated bytes, OR
 * allocating an extra array of 16 bytes and writing and copying from it

What's your preference? I don't have a good grasp of the performance difference between the two.

Reply:
Oops, pushed fix [3a7186b](https://github.com/elastic/elasticsearch/pull/97395/commits/3a7186b0572657d46dca352805a36e047f46d11b)

Reply:
> What's your preference? I don't have a good grasp of the performance difference between the two.

I've made a simple local test to compare the two implementations. 
It turns out that the second option (_allocating an extra array of 16 bytes and writing and copying from it_) is slightly faster. On average it's faster around 35ms than generating extra 8 bytes which are thrown away. I've tested it running `100 000` generations. With `10 000` generations, the difference is ~2-3ms.
Even though the difference is almost negligible, I'd prefer the second option.

Reply:
WOW, thanks for testing!!!

Reply:
Just to make it clear, `100 000` generations with option 1 is slower than option 2 by ~35ms. So for every `100 000` generations we save ~ 35ms. But what I did not consider is that this costs ~1.5 MB of RAM which has to be garbage collected. I have no idea how much time GC would take and it may be that in the end we break even.

Reply:
Right, it's tricky to benchmark durations that also involve the GC as the GC is influenced by the other activity in the VM.